include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
+	{
+		return bitCount_helper<const underlying_t>::__call(reinterpret_cast<const underlying_t&>(val));
+	}
+};
 
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int16_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint16_t, uint32_t)
-#ifndef __HLSL_VERSION
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int8_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint8_t, uint32_t)
-#endif
+template<typename T, typename U>
+requires concepts::FloatingPoint<T> && (concepts::FloatingPoint<T> || concepts::Boolean<T>)
+struct mix_helper<T, U>
+{
+	using return_t = T;
+	static inline return_t __call(const T x, const T y, const U a)
+	{
+		return glm::mix(x, y, a);
+	}
+};
 
-template<>
-struct find_lsb_helper<uint64_t>
+template<typename T>
+requires concepts::FloatingPointScalar<T> || concepts::IntegralScalar<T>
+struct sign_helper<T>
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(uint64_t) val)
+	using return_t = T;
+	static inline return_t __call(const T val)
 	{
-#ifdef __HLSL_VERSION
-		const uint32_t lowBits = uint32_t(val);
-		const int32_t lowLsb = find_lsb_helper<uint32_t>::findLSB(lowBits);
+		if (val < 0)
+			return -1;
+		if (val > 0)
+			return 1;
 
-		if (lowLsb == -1)
-		{
-			const uint32_t highBits = uint32_t(val >> 32);
-			const int32_t highLsb = find_lsb_helper<uint32_t>::findLSB(highBits);
-			if (highLsb == -1)
-				return -1;
-			else
-				return 32 + highLsb;
-		}
+		return 0;
+	}
+};
 
-		return lowLsb;
-#else
-		return glm::findLSB(val);
-#endif
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct radians_helper<T>
+{
+	using return_t = T;
+	static inline return_t __call(const T degrees)
+	{
+		return degrees * (numbers::pi<T> / static_cast<T>(180.0));
 	}
 };
 
-template<int N>
-struct find_lsb_helper<vector<int32_t, N> >
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct degrees_helper<T>
 {
-	static vector<int32_t, N> findLSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	using return_t = T;
+	static inline return_t __call(const T radians)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findILsb(val);
-#else
-		return glm::findLSB(val);
-#endif
+		return radians * (static_cast<T>(180.0) / numbers::pi<T>);
 	}
 };
 
-template<int N>
-struct find_lsb_helper<vector<uint32_t, N> >
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct step_helper<T>
 {
-	static vector<int32_t, N> findLSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	using return_t = T;
+	static inline return_t __call(const T edge, const T x)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findILsb(val);
-#else
-		return glm::findLSB(val);
-#endif
+		return x < edge ? 0.0 : 1.0;
 	}
 };
 
-#ifndef __HLSL_VERSION
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct smoothStep_helper<T>
+{
+	using return_t = T;
+	static inline return_t __call(const T edge0, const T edge1, const T x)
+	{
+		T t = clamp_helper<T>::__call((x - edge0) / (edge1 - edge0), 0, 1);
+		return t * t * (3 - 2 * t);
```

Comment:
all the literals need to be surrounded by `T()` currently you have INT-FLOAT resulting from `3-2*t`